src/main/java/fact/io/hdureader/BinTableReader.java

```diff
@@ -0,0 +1,149 @@
+package fact.io.hdureader;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * A reader for a FITS binary table.
+ *
+ * This class implements the iterable and iterator interfaces which allow for such nice things as
+ *
+ *        for(OptionalTypesMap p : binTableReader){
+ *              assertTrue(p.containsKey("Data"));
+ *         }
+ *
+ *
+ * The next() and the getNextRow() methods deliver a map with the data from one row in the table.
+ *
+ * Created by mackaiver on 18/11/16.
+ */
+public class BinTableReader implements Reader {
+
+        /**
+         * Check whether there is another row to return from this heap
+         * @return true if another row can be read.
+         */
+        @Override
+        public boolean hasNext() {
+            return numberOfRowsRead < numberOfRowsInTable;
+        }
+
+
+
+        private final DataInputStream stream;
+        private final List<BinTable.TableColumn> columns;
+        private int numberOfRowsRead = 0;
+        private final int numberOfRowsInTable;
+
+
+
+        private BinTableReader(BinTable binTable) {
+            this.stream = binTable.tableDataStream;
+            this.columns = binTable.columns;
+            this.numberOfRowsInTable = binTable.numberOfRowsInTable;
+        }
+
+        public static BinTableReader forBinTable(BinTable binTable){
+            return new BinTableReader(binTable);
+        }
+
+
+        public OptionalTypesMap<String, Serializable> getNextRow() throws IOException {
+            OptionalTypesMap<String, Serializable> map = new OptionalTypesMap<>();
+            for(BinTable.TableColumn c : columns){
+                if(c.repeatCount > 1){
+                    map.put(c.name, readArrayFromStream(c, stream));
+                } else if(c.repeatCount == 1) {
+                    map.put(c.name, readSingleValueFromStream(c, stream));
+                } else if(c.repeatCount == 0){
+                    map.put(c.name, null);
```

Comment:
We had some debate about this.
My first instinct was to not add it at all. Not sure if thats wise however.
Putting null at this point might not be too bad though. What the users sees in the end is the OptionalTypesMap which has getter methods for returning Optionals. At this point I hope NPEs are avoided.